cosmoz-sse.js

```diff
@@ -1,230 +1,87 @@
-// @license Copyright (C) 2019 Neovici AB - Apache 2 License
-import { PolymerElement } from '@polymer/polymer/polymer-element';
-
-const SSEHandlers = {
-	text: data => data,
-	json: data => JSON.parse(data)
-};
-
-/**
-	`<cosmoz-sse>` is an element that allows subscribing to server-sent events.
-
-	@customElement
-	@polymer
-	@demo demo/index.html
-*/
-class CosmozSSE extends PolymerElement {
-	static get is() {
-		return 'cosmoz-sse';
-	}
-
-	static get properties() {
-		return {
-			_configuration: {
-				type: Object,
-				computed: '_computeConfiguration(withCredentials)'
-			},
-			/**
-			 * The SSE endpoint.
-			 */
-			url: {
-				type: String
-			},
-
-			/**
-			 * The named events you wish to subscribe to.
-			 */
-			events: {
-				type: Array,
-				value: [],
-				observer: '_onEventsChange'
+// @license Copyright (C) 2020 Neovici AB - Apache 2 License
+import {
+	component, useEffect, useMemo, useState
+} from 'haunted';
+
+const getJson = input => {
+		try {
+			return JSON.parse(input);
+		} catch (e) {
+			// eslint-disable-next-line no-console
+			console.error('cant parse JSON:', input);
+		}
+	},
+	SSEHandlers = {
+		text: data => data,
+		json: data => getJson(data)
+	},
+
+	// eslint-disable-next-line max-lines-per-function
+	CosmozSSE = function ({
+		events, withCredentials = false, handleAs = 'json', url
+	}) {
+		const fireEvent = (type, eventInit) => {
+				this.dispatchEvent(new CustomEvent(type, eventInit));
 			},
-
-			/**
-			 * How to parse the event data.
-			 * @type {('json'|'text')}
-			 */
-			handleAs: {
-				type: String,
-				value: 'json'
+			configuration = useMemo(() => ({
+				withCredentials
+			}), [withCredentials]),
+			[source, setSource] = useState(null),
+			eventTypes = useMemo(() => events && getJson(events) || [], [events]),
+			onEvent = event => {
+				const handler = SSEHandlers[handleAs];
+
+				fireEvent(event.type, {
+					detail: {
+						data: handler(event.data)
+					}
+				});
 			},
-			/**
-			 * Whether CORS should include credentials.
-			 */
-			withCredentials: {
-				type: Boolean,
-				value: false
+			unsubscribeFromEvents = () => {
+				if (source == null || !Array.isArray(eventTypes)) {
+					return;
+				}
+				eventTypes.forEach(eventType =>
+					source.removeEventListener(eventType, onEvent)
+				);
+			};
```

Comment:
Normally you should use `useCallback` to define these, but in this case you should move these two callbacks in the effect that does the `events` subscriptions. This way all relevant code will be contained in the effect.

Reply:
Well previous behavior was to unsubscribe on teardown as well. Events won't have changed so it will be torn down without unsubscribing first?

Reply:
> Well previous behavior was to unsubscribe on teardown as well. Events won't have changed so it will be torn down without unsubscribing first?
You would set `events` to be a dependency of useCallback, so a new callback is created whenever events change. But this is needlessly complex. A better route would be to define the callbacks in the effect.

The behavior wouldn't change if you define `onEvent` and `unsubscribeFromEvents` in the effect. It would help contain the implementation details. There is no need to have these two functions defined outside of the effect, because you do not use them elsewhere.
Just like you did with the effect that creates the EventSource.

Reply:
`unsubscribeFromEvents` is used in two places, that's what I mean?
If `url` changes, the teardown of the EventSource will happen without events teardown happening?

Reply:
If `url` changes, then `source` changes, then the old events are unsubscribed.